File: Interpreter/src/parser.zig
const std = @import("std");
const t = @import("token.zig");
const l = @import("lexer.zig");
const a = @import("ast.zig");

const OperatorPrecidence = enum(u16) {
    lowest = 1,
    equals,
    less_than,
    greater_than,
    sum,
    product,
    prefix,
    infix,
};

pub const Parser = struct {
    lexer: *l.Lexer,
    cur_token: t.Token,
    next_token: t.Token,
    ast: std.ArrayList(a.Node),
    parsed_tokens_arena: std.heap.ArenaAllocator,
    parsed_tokens_alloc: std.mem.Allocator,
    errors: std.ArrayList([]const u8),
    precidenceMap: std.AutoHashMap(t.TokenType, OperatorPrecidence),

    pub fn init(alloc: std.mem.Allocator, lex: *l.Lexer) Parser {
        var p = Parser{
            .lexer = lex,
            .cur_token = undefined,
            .next_token = undefined,
            .ast = std.ArrayList(a.Node).init(alloc),
            .parsed_tokens_arena = std.heap.ArenaAllocator.init(alloc),
            .parsed_tokens_alloc = undefined,
            .errors = std.ArrayList([]const u8).init(alloc),
            .precidenceMap = std.AutoHashMap(t.TokenType, OperatorPrecidence).init(alloc),
        };

        p.parsed_tokens_alloc = p.parsed_tokens_arena.allocator();

        p.precidenceMap.put(t.TokenType.equal, OperatorPrecidence.equals) catch unreachable;
        p.precidenceMap.put(t.TokenType.not_equal, OperatorPrecidence.equals) catch unreachable;
        p.precidenceMap.put(t.TokenType.less_than, OperatorPrecidence.less_than) catch unreachable;
        p.precidenceMap.put(t.TokenType.greater_than, OperatorPrecidence.less_than) catch unreachable;
        p.precidenceMap.put(t.TokenType.plus, OperatorPrecidence.sum) catch unreachable;
        p.precidenceMap.put(t.TokenType.minus, OperatorPrecidence.sum) catch unreachable;
        p.precidenceMap.put(t.TokenType.divide, OperatorPrecidence.product) catch unreachable;
        p.precidenceMap.put(t.TokenType.multiply, OperatorPrecidence.product) catch unreachable;

        p.increment();
        p.increment();
        return p;
    }
    pub fn deinit(self: *Parser) void {
        self.ast.deinit();
        self.errors.deinit();
        self.parsed_tokens_arena.deinit();
        self.precidenceMap.deinit();
    }

    fn increment(self: *Parser) void {
        self.cur_token = self.next_token;
        self.next_token = self.lexer.nextToken();
    }

    fn assert(self: *Parser, expected: t.TokenType, err: []const u8) bool {
        if (self.cur_token.type == expected) {
            return true;
        }
        self.errors.append(err) catch unreachable;
        return false;
    }

    fn assertPeek(self: *Parser, expected: t.TokenType, err: []const u8) bool {
        if (self.next_token.type == expected) {
            self.increment();
            return true;
        }
        self.errors.append(err) catch unreachable;
        return false;
    }

    fn peekPrecidence(self: *Parser) OperatorPrecidence {
        if (self.precidenceMap.get(self.next_token.type)) |p| {
            return p;
        }
        return OperatorPrecidence.lowest;
    }

    fn curPrecidence(self: *Parser) OperatorPrecidence {
        if (self.precidenceMap.get(self.cur_token.type)) |p| {
            return p;
        }
        return OperatorPrecidence.lowest;
    }

    pub fn parseTokens(self: *Parser) void {
        while (self.cur_token.type != t.TokenType.eof) {
            const node = switch (self.cur_token.type) {
                t.TokenType.let => self.parseLetToken(),
                t.TokenType._return => self.parseReturnToken(),
                else => self.parseExpressionStatement(),
            };

            if (node) |n| {
                self.ast.append(n) catch unreachable;
            }
            self.increment();
        }
    }

    // Parsing Expressions {{{

    fn parseExpression(self: *Parser, precidence: OperatorPrecidence) ?a.Expression {
        var left_expr = switch (self.cur_token.type) {
            t.TokenType.identifier => self.parseIdentifier(),
            t.TokenType.int => self.parseIntLiteral(),
            t.TokenType.minus => self.parsePrefixExpression(),
            t.TokenType.bang => self.parsePrefixExpression(),

            else => null,
        };
        if (left_expr == null) return null;

        while (self.next_token.type != t.TokenType.semicolon and
            @intFromEnum(precidence) < @intFromEnum(self.peekPrecidence()))
        {
            self.increment();
            left_expr = switch (self.cur_token.type) {
                t.TokenType.plus => self.parseInfixExpression(left_expr),
                t.TokenType.minus => self.parseInfixExpression(left_expr),
                t.TokenType.multiply => self.parseInfixExpression(left_expr),
                t.TokenType.divide => self.parseInfixExpression(left_expr),
                t.TokenType.less_than => self.parseInfixExpression(left_expr),
                t.TokenType.greater_than => self.parseInfixExpression(left_expr),
                t.TokenType.equal => self.parseInfixExpression(left_expr),
                t.TokenType.not_equal => self.parseInfixExpression(left_expr),
                else => null,
            };
            if (left_expr == null) {
                return null;
            }
        }
        return left_expr;
    }

    fn parseInfixExpression(self: *Parser, left: ?a.Expression) ?a.Expression {
        if (left == null) {
            return null;
        }

        var expr = a.InfixExpression{
            .tokenLiteral = self.cur_token,
            .opperator = self.cur_token.literal,
            .left = left.?,
            .right = undefined,
        };

        const precidence = self.curPrecidence();
        self.increment();
        if (self.parseExpression(precidence)) |e| {
            expr.right = e;
            return expr.expression();
        } else {
            self.errors.append("unable to parse right side to infix") catch unreachable;
            return null;
        }
    }

    fn parsePrefixExpression(self: *Parser) ?a.Expression {
        var exp = self.parsed_tokens_alloc.create(a.PrefixExpression) catch unreachable;
        exp.* = a.PrefixExpression{
            .tokenLiteral = self.cur_token,
            .opperator = self.cur_token.literal,
            .right = undefined,
        };
        self.increment();

        // parse right side to expression
        if (self.parseExpression(OperatorPrecidence.prefix)) |e| {
            exp.right = e;
        } else {
            self.errors.append("no expression after prefix") catch unreachable;
            return null;
        }

        return exp.expression();
    }

    fn parseIntLiteral(self: *Parser) a.Expression {
        var expr = self.parsed_tokens_alloc.create(a.IntLiteral) catch unreachable;
        expr.* = a.IntLiteral{
            .tokenLiteral = self.cur_token,
            .value = std.fmt.parseInt(i64, self.cur_token.literal, '0') catch unreachable,
        };
        return expr.expression();
    }

    fn parseIdentifier(self: *Parser) a.Expression {
        var expr = self.parsed_tokens_alloc.create(a.Identifier) catch unreachable;
        expr.* = a.Identifier{
            .tokenLiteral = self.cur_token,
            .value = self.cur_token.literal,
        };
        return expr.expression();
    }
    // }}}

    // Parsing Statements {{{
    fn parseReturnToken(self: *Parser) ?a.Node {
        var stmt = self.parsed_tokens_alloc.create(a.ReturnStatement) catch unreachable;
        stmt.* = a.ReturnStatement{
            .tokenLiteral = self.cur_token,
            .value = undefined,
        };
        stmt.value = a.Expression{
            .ptr = undefined,
            .toStringFn = undefined,
        };

        while (self.cur_token.type != t.TokenType.semicolon) {
            self.increment();
        }

        return stmt.node();
    }

    fn parseLetToken(self: *Parser) ?a.Node {
        var stmt = self.parsed_tokens_alloc.create(a.LetStatement) catch unreachable;
        stmt.* = a.LetStatement{
            .tokenLiteral = self.cur_token,
            .ident = undefined,
            .value = undefined,
        };

        // parse identifier
        if (!self.assertPeek(t.TokenType.identifier, "Expected Identifier")) {
            return null;
        }
        stmt.ident = self.cur_token.literal;

        // check assign
        if (!self.assertPeek(t.TokenType.assign, "Expected '='")) {
            return null;
        } else {
            self.increment();
        }

        // parse expression
        stmt.value = a.Expression{
            .ptr = undefined,
            .toStringFn = undefined,
        };

        while (self.cur_token.type != t.TokenType.semicolon) {
            self.increment();
        }

        // check semicolon
        if (!self.assert(t.TokenType.semicolon, "Expected ';'")) {
            return null;
        }

        return stmt.node();
    }

    fn parseExpressionStatement(self: *Parser) ?a.Node {
        var stmt = self.parsed_tokens_alloc.create(a.ExpressionStatement) catch unreachable;
        stmt.* = a.ExpressionStatement{
            .tokenLiteral = self.cur_token,
            .value = undefined,
        };

        // parse expression
        if (self.parseExpression(OperatorPrecidence.lowest)) |val| {
            stmt.value = val;
        } else {
            self.errors.append("invalid expression") catch unreachable;
            return null;
        }

        if (self.next_token.type == t.TokenType.semicolon) {
            self.increment();
        }

        // return interface
        return stmt.node();
    }
    // }}}
};

File: Interpreter/go-impl/parser/parser.go
package parser

import (
	"interpreter/ast"
	"interpreter/lexer"
	"interpreter/token"
	"strconv"
)

const (
    precidence_Lowest = iota
    precidence_Equals
    precidence_LessThan
    precidence_GreaterThan
    precidence_Sum
    precidence_Product
    precidence_Prefix
    precidence_Infix
);

var precidenceMap = map[uint32]int {
    token.Op_equal: precidence_Equals,
    token.Op_notEqual: precidence_Equals,
    token.Op_lessthan: precidence_LessThan,
    token.Op_greaterthan: precidence_GreaterThan,
    token.Op_plus: precidence_Sum,
    token.Op_minus: precidence_Sum,
    token.Op_asterisk: precidence_Product,
    token.Op_slash: precidence_Product,

}

type Parser struct {
    lex *lexer.Lexer       
    curToken token.Token
    nextToken token.Token
    ast []ast.Statement
    errors []string
}

func New(lex *lexer.Lexer) Parser {
    p := Parser {
        lex: lex,
    }
    p.Incr()
    p.Incr()
    return p
}

func (p *Parser) ParseTokens() {
    for p.curToken.TokenType != token.Eof {
        if node := p.parse(); node != nil {
            p.ast = append(p.ast, node)
        }
        p.Incr()
    }
}

func (p *Parser) parse() ast.Statement {
    switch p.curToken.TokenType {
        case token.Keyw_let: return p.parseLetStatement()
        case token.Keyw_return: return p.parseReturnStatement()
        default: return p.parseExpressionStatement()
    }
}

// parse statements {{{
func (p *Parser) parseLetStatement() ast.Statement {
    stmt := ast.LetStatement {
        Token: p.curToken,
    }
    p.Incr()

    // parse identifier
    stmt.Identifier = p.parseExpression(precidence_Lowest)
    if stmt.Identifier == nil {
        return nil
    } else {
        p.Incr()
    }

    // check syntax
    if !p.assert(token.Syn_assign, "invalid syntax: expected '='") {
        return nil
    }
    p.Incr()

    // parse expression
    stmt.Value = p.parseExpression(precidence_Lowest)
    if stmt.Value == nil {
        return nil
    } else {
        p.Incr()
    }

    // check syntax
    if !p.assert(token.Syn_semicolon, "invalid syntax: expected ';'") {
        return nil
    }
    return &stmt
}

func (p *Parser) parseReturnStatement() ast.Statement {
    stmt := ast.ReturnStatement {
        Token: p.curToken,
    }
    p.Incr()

    stmt.Value = p.parseExpression(precidence_Lowest)
    if stmt.Value == nil {
        return nil
    } else {
        p.Incr()
    }

    // check syntax
    if !p.assert(token.Syn_semicolon, "invalid syntax: expected ';'") {
        return nil
    }
    return &stmt
}

func (p *Parser) parseExpressionStatement() ast.Statement {
    stmt := ast.ExpressionStatement {
        Token: p.curToken,
    }

    stmt.Value = p.parseExpression(precidence_Lowest)
    if stmt.Value == nil {
        p.errors = append(p.errors, "could not parse expression")
        return nil
    }
    p.Incr()

    if !p.assert(token.Syn_semicolon, "invalid syntax: expected ';'") {
        return nil
    }
    return &stmt
}
// }}}

var indent int = 0
// parse expressions {{{
func (p *Parser) parseExpression(precidence int) ast.Expression {
    var leftExpr ast.Expression
    switch p.curToken.TokenType {
        case token.Type_identifier: leftExpr = p.parseIdentExpression()
        case token.Type_int: leftExpr = p.parseIntLiteral()
        case token.Type_bool: leftExpr = p.parseBoolLiteral()
        case token.Op_bang: leftExpr = p.parsePrefixExpression()
        case token.Op_minus: leftExpr = p.parsePrefixExpression()
        case token.Syn_lparen: leftExpr = p.parseParenExpr()
        default: leftExpr = nil
    }
    if leftExpr == nil {
        return nil
    }

    for p.curToken.TokenType != token.Eof && precidence < p.getPrecidence(p.nextToken.TokenType) {
        p.Incr()
        switch p.curToken.TokenType {
            case token.Op_plus: leftExpr = p.parseInfixExpression(leftExpr)
            case token.Op_minus: leftExpr = p.parseInfixExpression(leftExpr)
            case token.Op_asterisk: leftExpr = p.parseInfixExpression(leftExpr)
            case token.Op_slash: leftExpr = p.parseInfixExpression(leftExpr)
            case token.Op_lessthan: leftExpr = p.parseInfixExpression(leftExpr)
            case token.Op_greaterthan: leftExpr = p.parseInfixExpression(leftExpr)
            case token.Op_equal: leftExpr = p.parseInfixExpression(leftExpr)
            case token.Op_notEqual: leftExpr = p.parseInfixExpression(leftExpr)
            default: leftExpr = nil
        }
        if leftExpr == nil {
            return nil
        }

    }
    return leftExpr
}

func (p *Parser) parseIdentExpression() ast.Expression {
    return &ast.Identifier {
        Token: p.curToken,
        Value: p.curToken.Literal,
    }
}

func (p *Parser) parseIntLiteral() ast.Expression {
    expr := &ast.IntLiteral{
        Token: p.curToken,
    }
    if v,err := strconv.Atoi(p.curToken.Literal); err == nil {
        expr.Value = int64(v)
    } else {
        p.errors = append(p.errors, "Invalid int literal")
    }
    return expr
}

func (p *Parser) parseInfixExpression(left ast.Expression) ast.Expression {
    if left == nil {
        return nil
    }

    expr := ast.InfixExpression {
        Token: p.curToken,
        Opperator: p.curToken.Literal,
        Left: left,
    }
    prec := p.getPrecidence(p.curToken.TokenType)
    p.Incr()

    expr.Right = p.parseExpression(prec)
    if expr.Right == nil {
        return nil
    }
    return &expr
}

func (p *Parser) parsePrefixExpression() ast.Expression {
    expr := ast.PrefixExpression {
        Token: p.curToken,
        Opperator: p.curToken.Literal,
    }
    p.Incr()

    expr.Right = p.parseExpression(precidence_Prefix)
    if expr.Right == nil {
        return nil
    }
    return &expr
}

func (p *Parser) parseBoolLiteral() ast.Expression {
    expr := ast.BoolLiteral {
        Token: p.curToken,
    }
    b,err := strconv.ParseBool(p.curToken.Literal)
    if err != nil {
        p.errors = append(p.errors, "Invalid bool literal")
        return nil
    }
    expr.Value = b
    return &expr
}

func (p *Parser) parseParenExpr() ast.Expression {
    p.Incr()
    expr := p.parseExpression(precidence_Lowest)

    if expr == nil {
        p.errors = append(p.errors, "could not parse expression in parens")
        return nil
    }
    if p.nextToken.TokenType != token.Syn_rparen {
        p.errors = append(p.errors, "invalid syntax: expected ')'")
        return nil
    }
    p.Incr()
    return expr
}
//}}}



func (p *Parser) Incr() {
    p.curToken = p.nextToken
    p.nextToken = p.lex.NextToken()
}

func (p *Parser) assert(expected uint32, err string) bool {
    if p.curToken.TokenType == expected {
        return true
    }
    p.errors = append(p.errors, err)
    return false
}

func (p *Parser) getPrecidence(tok uint32) int {
    if pres,ok := precidenceMap[tok]; ok {
        return pres 
    }
    return precidence_Lowest
}
